File: kaleido-gradle-plugin/src/main/java/com/tongsr/kaleido/gradle/dsl/KaleidoGeneration.java
package com.tongsr.kaleido.gradle.dsl;

import javax.inject.Inject;
import org.gradle.api.Action;
import org.gradle.api.model.ObjectFactory;
import org.gradle.api.provider.Property;

public abstract class KaleidoGeneration {
    private final KaleidoComposeGeneration compose;

    @Inject
    public KaleidoGeneration(ObjectFactory objects) {
        compose = objects.newInstance(KaleidoComposeGeneration.class);
    }

    public abstract Property<String> getPackageBase();

    public abstract Property<Integer> getPackageCount();

    public abstract Property<Integer> getClassesPerPackage();

    public abstract Property<Integer> getMethodsPerClass();

    public abstract Property<Integer> getLayoutCount();

    public abstract Property<Integer> getDrawableCount();

    public abstract Property<Integer> getStringCount();

    public abstract Property<Integer> getActivityCount();

    public KaleidoComposeGeneration getCompose() {
        return compose;
    }

    public void compose(Action<? super KaleidoComposeGeneration> action) {
        action.execute(compose);
    }
}

File: kaleido-gradle-plugin/src/main/kotlin/com/tongsr/kaleido/gradle/dsl/KaleidoGeneration.kt
package com.tongsr.kaleido.gradle.dsl

import org.gradle.api.Action
import org.gradle.api.model.ObjectFactory
import org.gradle.api.provider.Property
import javax.inject.Inject

abstract class KaleidoGeneration @Inject constructor(objects: ObjectFactory) {
    abstract val packageBase: Property<String>
    abstract val packageCount: Property<Int>
    abstract val classesPerPackage: Property<Int>
    abstract val methodsPerClass: Property<Int>
    abstract val layoutCount: Property<Int>
    abstract val drawableCount: Property<Int>
    abstract val stringCount: Property<Int>
    abstract val activityCount: Property<Int>

    val compose: KaleidoComposeGeneration = objects.newInstance(KaleidoComposeGeneration::class.java)

    fun compose(action: Action<in KaleidoComposeGeneration>) {
        action.execute(compose)
    }
}
